File: Selection.js
var words;
var r; // min length of word
var separator = ", ";
var i;
var complete_group = [];


function groupSelection(candidates, size) {
    words = candidates;
    r = size;

    i = 0;
    recursiveFillUp("");
    return complete_group;
}

// pick elements recursively
function recursiveFillUp(temp) {
    var picked_elements = [];
    var j = i;
    while (j < words.length) {
        picked_elements[j] = temp + words[j];
        // recoil factor
        if (i >= words.length) {
            i = j;
        }
        // satisfied yet?
        if (picked_elements[j].split(separator).length == r) {
            complete_group.push(picked_elements[j]);
        } else if (picked_elements[j].split(separator).length < r) {
            recursiveFillUp(picked_elements[j] + separator);
        }
        j++;
    }
    if (picked_elements[--j] !== undefined &&
            picked_elements[j].split(separator).length == r) {
        i++; // keep recoil factor straightened out
    }
}

File: Selection.java
package miscellaneous;

import java.util.ArrayList;
import java.util.List;

public class Selection {

    public String[] words;
    public int r; // min length of word
    protected List<String[]> complete_group;
    private int i;

    public Selection() {
    }

    public List<String[]> groupSelection(String[] candidates, int size) {
        words = candidates;
        r = size;
        complete_group = new ArrayList<>();
        i = 0;
        recursiveFillUp(new ArrayList<>());

        return complete_group;
    }

// pick elements recursively
    protected void recursiveFillUp(List<String> temp) {
        List<String>[] picked_elements = new List[words.length];
        int j = i;
        while (j < words.length) {
            picked_elements[j] = new ArrayList<>();
            picked_elements[j].addAll(temp);
            picked_elements[j].add(words[j]);
            // recoil factor
            if (i >= words.length) {
                i = j;
            }
            // satisfied yet?
            if (picked_elements[j].size() == r) {
                complete_group.add(picked_elements[j].toArray(new String[0]));
            } else if (picked_elements[j].size() < r) {
                recursiveFillUp(picked_elements[j]);
            }
            j++;
        }
        if (picked_elements[--j] != null && picked_elements[j].size() == r) {
            i++; // keep recoil factor straightened out
        }
    }
}
